relay.go

```diff
@@ -50,26 +52,36 @@ func (r *Relayer) Hosts() RelayHosts {
 
 // Relay forwards a frame.
 func (r *Relayer) Relay(f *Frame) error {
-	// TODO: remove relay items on timeout.
 	if f.messageType() != messageTypeCallReq {
-		r.RLock()
-		item, ok := r.items[f.Header.ID]
-		r.RUnlock()
-		if !ok {
-			return errors.New("non-callReq for inactive ID")
-		}
-		f.Header.ID = item.remapID
-		item.destination.Receive(f)
-		if f.isLast() {
-			r.removeRelayItem(f.Header.ID)
-		}
-		return nil
+		return r.handleNonCallReq(f)
 	}
+	return r.handleCallReq(f)
+}
 
-	// Handle messageTypeCallReq
+// Receive accepts a relayed frame.
+func (r *Relayer) Receive(f *Frame) {
+	r.RLock()
+	_, ok := r.items[f.Header.ID]
+	r.RUnlock()
+	if !ok {
+		r.logger.WithFields(
+			LogField{"ID", f.Header.ID},
+		).Warn("Received a frame without a RelayItem.")
+	}
+
+	r.conn.sendCh <- f
+	if finishesCall(f) {
+		r.removeRelayItem(f.Header.ID)
+	}
+}
+
+func (r *Relayer) handleCallReq(f *Frame) error {
+	r.RLock()
 	if _, ok := r.items[f.Header.ID]; ok {
+		r.RUnlock()
```

Comment:
while I normally try to scope variables in the `if`, I'd prefer to avoid multiple `Unlock` calls. Suggestions on how to fix:
1. 
```go
r.RLock()
_, ok := r.items[f.Header.ID]
r.RUnlock()
if !ok {
  [...]
}
```

2. Break out into a function `func (r *Relayer) hasRelay(id uint32) bool` which gets the `RLock`